iOS support.

VladasZ opened the pull request:
Added iOS support. Also I deleted and gitignored `src/bindings.rs` file because it is generated and shouldn't be tracked. Feel free to correct my grammar since English is not my native language.

nsaslavsky commented on build.rs:
```diff
@@ -1,27 +1,46 @@
 
+    // println!("{:?}", ndk_include_dir);
+    println!("cargo:rustc-link-lib=log");
+    println!("cargo:rustc-link-lib=android");
```
This crate is just for opengles bindings it should not link against `liblog`, `libandroid` and `libegl` from the NDK. If an application wants to link against it should go in apps build.rs or come from a different crate.

VladasZ replied:
Fixed

nsaslavsky commented on build.rs:
```diff
@@ -1,27 +1,46 @@
+}
+
+fn main() {
+    if matches!(env::var("ANDROID_BUILD"), Ok(_)) {
```
Instead of this non-standard environment variable please use something like `std::env::var("CARGO_CFG_TARGET_OS") == Ok("android".into()`

VladasZ replied:
Fixed

nsaslavsky commented on build.rs:
```diff
@@ -1,27 +1,46 @@
-    let gl31h = ndk_include_dir.join("GLES3/gl31.h");
-    let gl31h = gl31h.to_str().unwrap();
-    let ndk_include_dir = ndk_include_dir.to_str().unwrap();
+fn android_setup() {
```
I don't like that you commented out all code for generating the bindings, but I understand that it unreasonable to expect a user targeting iOS to have the NDK installed. I think that a better approach would be to include a small bash script to run bindgen and regenerate the headers. Something like what has been done for [ndk-sys](https://github.com/rust-windowing/android-ndk-rs/tree/master/ndk-sys) would be good.

VladasZ replied:
Fixed

avantgardnerio commented:
Thank you for this PR, and I'm very sorry that no one has reviewed it in many months. I've added a codeowners file that will ensure our organization is diligent with code reviews in the future.

Due to the complexities of maintaining a FFI binding that works on multiple operating systems, I think CI is required in order to ensure that all targets remain up to date. I've created a github action that ensures the Android build continues to work, as well as stubbing out a MacOS version.

If you can update the MacOS build to complete successfully, then we will review this PR and if it doesn't break our build we will merge it.

I hope we can work together to create 1 crate that works for everyone, even if that causes us both a little more work.

VladasZ commented:
@nsaslavsky thanks for your review. I opened this PR 9 months ago when I just started learning Rust and barely understood what is `bingen` and how `build.rs` works. I will draft it, then completely rewrite and when it is ready I will re-request review from you again.

avantgardnerio commented:
> @nsaslavsky thanks for your review. I opened this PR 9 months ago when I just started learning Rust and barely understood what is `bingen` and how `build.rs` works. I will draft it, then completely rewrite and when it is ready I will re-request review from you again.

No worries! We're all learning. The project this was for was all of our first production Rust code I think :)

VladasZ committed:
iOS support


install lipo

generate bindings func


proper iOS/Android check


Fixed android linker error


iOS bindgen


Build stage


Delete and ignore bindings


Check if symlink exists


Android linker explanation


Reworked iOS includes


Updated readme

nsaslavsky commented on build.rs:
```diff
@@ -1,22 +1,62 @@
+    }
+
+    fs::create_dir("temp").unwrap();
+    std::os::unix::fs::symlink(include_dir, gles_dir).unwrap();
```
This will cause a compile error on windows, maybe add something like `#[cfg(target_os = "macos")]`.

VladasZ replied:
Fixed

nsaslavsky reviewed (commented):
Sorry, I took a while to review this, it just needs one more small change.

VladasZ committed:
Windows fix

avantgardnerio commented:
Nice work, thank you @VladasZ !

VladasZ commented:
@avantgardnerio @nsaslavsky thanks for approving. Can you release it?

avantgardnerio commented:
Sorry, I forgot, we will do this shortly.

avantgardnerio commented:
Done: https://crates.io/crates/gles31-sys/versions
